### PR TITLE
nixos/regreet: init

### PR DESCRIPTION
**Pull request by fufexan**
###### Description of changes

- ReGreet: 2023-02-27 -> 2023-03-12
- nixos/regreet: init

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by spikespaz** [comment]
`extraCss` could be `types.either types.path types.lines`

**Comment by spikespaz** [comment]
Might want to leave a note somewhere that this does not automatically work, and just enables `services.greetd.enable`. Does `services.greetd` config have some kind of assert to make sure `services.greetd.settings.default_session` exists? If not, you might want to add an assert here, with a message saying something like:
> You must set (for example)
> ```
> services.greetd.settings.default_session.command = lib.escapeShellArgs [
>   (lib.getExe config.programs.regreet.package)
> ];
> ```

Or otherwise, provide an RO option which contains the command-line determined by options of this module, or have it (optionally) directly set the command for `default_session`.

Several approaches may be appropriate, some probably more than others, but as it stands if someone tries to use this module (an they are new to this universe) they will have many questions.

File: nixos/modules/programs/regreet.nix [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+      '';
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    services.greetd.enable = true;
+
+    environment.etc = {
+      "greetd/regreet.toml".source = settingsType.generate "regreet.toml" cfg.settings;
+      "greetd/regreet.css".source = pkgs.writeText "regreet.css" cfg.extraCss;
```
Comment (spikespaz) [inline nixos/modules/programs/regreet.nix:42]: Does `pkgs.writeText` allow for a `lib.types.path` to be passed as a second parameter? I was expecting to see this line changed to
```nix
{
  "greetd/regreet.css" =
    if lib.isPath cfg.extraCss
    then {source = cfg.extraCss;}
    else {text = cfg.extraCss;};
}
```

Comment (spikespaz) [reply nixos/modules/programs/regreet.nix:42]: It is possible that users may also want to do something similar for `regreet.toml`, however I would consider this "discouraged" since there is a `pkgs.formats.toml` to use. CSS is a different story for obvious reasons.

File: nixos/modules/programs/regreet.nix [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    };
+
+    systemd.tmpfiles.rules = [
+      "d /var/log/regreet 0755 greeter greeter - -"
```
Comment (spikespaz) [inline nixos/modules/programs/regreet.nix:46]: Since `services.greetd.settings.<session>.user` is an option that does exist, perhaps there should be another option exposed for this module, `user`.

File: nixos/modules/programs/regreet.nix [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+  settingsType = pkgs.formats.toml { };
+in
+{
+  options.programs.regreet = {
+    enable = lib.mkEnableOption (lib.mdDoc "ReGreet, a greetd greeter");
+
+    package = lib.mkPackageOptionMD pkgs "regreet" { };
+
+    settings = lib.mkOption {
+      inherit (settingsType) type;
```
Comment (spikespaz) [inline nixos/modules/programs/regreet.nix:17]: Just a nitpick: I don't like the way this reads: `type (dot) type`. I would prefer `format (dot) type`. Indeed, I see many other modules using `settingsFormat` or `tomlFormat`, I personally would choose `tomlFormat` but I believe this is less common (and more unconventional) than the former.

**Comment by fufexan** [comment]
> You must set (for example)
> 
> ```
> services.greetd.settings.default_session.command = lib.escapeShellArgs [
>   (lib.getExe config.programs.regreet.package)
> ];
> ```

That would work if it was a standalone program capable of displaying itself. But it is just a GTK program that can run as an X window or a wayland layer shell. So there would need to be a way to either specify which WM/compositor to run regreet with (at which point why not just instruct the user to write the default_command themselves?), or use a default which is known to work everywhere, and is small (maybe weston or cage).

Opinions on this are welcome.

File: nixos/modules/programs/regreet.nix (hunk 1 above)
Comment (fufexan) [reply nixos/modules/programs/regreet.nix:42]: Sure, I'll try to modify the functionality of the toml settings as well.



> Does `pkgs.writeText` allow for a `lib.types.path` to be passed as a second parameter? I was expecting to see this line changed to

I forgot :sweat_smile:

File: nixos/modules/programs/regreet.nix (hunk 2 above)
Comment (fufexan) [reply nixos/modules/programs/regreet.nix:46]: Indeed, I haven't thought of this. Good point.

**Comment by spikespaz** [comment]
> That would work if it was a standalone program capable of displaying itself

Ah, yes, I missed that. I mistakenly referenced my own config using `tuigreet` when writing that example.

With this in mind, I would like to move the discussion to an adjacent issue:

The purpose of `services.greetd.enable` when this module otherwise affects the behavior of `greetd` in no way. As you say yourself, ReGreet is standalone, and theoretically someone in the world may want to reset `greetd.enable = false` for some reason.

Since getting this module to work in tandem with `greetd` requires much more explicit configuration, I would vote to remove the configuration line, as it serves no purpose (?) without more elaboration.

The user of this module must opt-in to usage of `greetd`, regardless of `services.greetd.enable` being set. We may as well make this requirement explicit.

**Comment by spikespaz** [comment]
Alternatively, the scope of this module could be expanded to provide an out-of-the-box working configuration for "compositor + greeter + daemon", but I believe it would be difficult to make the configuration options properly generic over all possible desired combinations. This may not be a concern however, as the NixOS contributor guide says to only provide the most common desired options for the most minimal configuration possible. Then that leads us back to just not addressing the issue in the first place.

**Comment by fufexan** [comment]
I'm in favour of that, but then what do we set as the user of the tmpfiles?

If the user sets a different `user` for a specific session, then `greeter` will not match anymore and ReGreet will fail to write its files there.
I think it's safe to use `config.services.greetd.settings.default_session.user` though.

**Comment by spikespaz** [comment]
> I'm in favour of that, but then what do we set as the user of the tmpfiles?
> 
> If the user sets a different `user` for a specific session, then `greeter` will not match anymore and ReGreet will fail to write its files there.
> I think it's safe to use `config.services.greetd.settings.default_session.user` though.

I would agree. Check that the `default_session.user` attribute path exists, if it does not, use a default. I need to see the greetd module in a moment to determine how they handle the user for the daemon though, before I'm certain about this approach.

**Comment by fufexan** [comment]
https://github.com/NixOS/nixpkgs/blob/9e91539bb454af74b2e8f012e1aae08cb8abd9ff/nixos/modules/services/display-managers/greetd.nix#L57

We can use it as-is.

File: nixos/modules/programs/regreet.nix (hunk 3 above)
Comment (spikespaz) [reply nixos/modules/programs/regreet.nix]: <https://github.com/NixOS/nixpkgs/blob/9e91539bb454af74b2e8f012e1aae08cb8abd9ff/nixos/modules/services/display-managers/greetd.nix#L7>

**Comment by spikespaz** [comment]
I'm questioning myself now. Should the ownership of ReGreet's logs respect the user that `regreet` was started as, or should it respect the `user` greeter which always exists?

<https://github.com/NixOS/nixpkgs/blob/9e91539bb454af74b2e8f012e1aae08cb8abd9ff/nixos/modules/services/display-managers/greetd.nix#L104>

I'm guessing that no matter who runs `regreet`, the human would probably expect the logs to be readable by those of the `greeter` group. Perhaps this would mean that different user/group would be appropriate?

File: nixos/modules/programs/regreet.nix [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+  config = lib.mkIf cfg.enable {
+    environment.etc = {
+      "greetd/regreet.toml" =
+        if lib.isPath cfg.extraCss
+        then {source = cfg.extraCss;}
+        else {text = cfg.extraCss;};
+
+      "greetd/regreet.css".source =
+        if lib.isPath cfg.settings
```
Comment (spikespaz) [inline nixos/modules/programs/regreet.nix:55]: These two are reversed.

File: nixos/modules/programs/regreet.nix [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        then settingsFormat.generate "regreet.toml" cfg.settings
+        else cfg.settings;
+    };
+
+    systemd.tmpfiles.rules = let
+      user = config.services.greetd.settings.default_session.user;
+    in [
+      "d /var/log/regreet 0755 ${user} ${user} - -"
```
Comment (spikespaz) [inline nixos/modules/programs/regreet.nix:63]: Adding on to my comment in the main chain,
`"d /var/log/regreet 0755 greeter ${user}"` should be more appropriate (I think). Not that it really matters, since public permissions are `xr-`.

File: nixos/modules/programs/regreet.nix (hunk 4 above)
Comment (fufexan) [reply nixos/modules/programs/regreet.nix:55]: My bad.

File: nixos/modules/programs/regreet.nix [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+, config
+, ...
+}:
+let
+  cfg = config.programs.regreet;
+  settingsFormat = pkgs.formats.toml { };
+in
+{
+  options.programs.regreet = {
+    enable = lib.mkEnableOption (lib.mdDoc "") // {
```
Comment (spikespaz) [inline nixos/modules/programs/regreet.nix:12]: What is the `//` for? If this is a mistake: value of `description` is just a string, not a markdown doc-string. The first parameter of `lib.mkEnableOption` is `lib.mdDoc ""` which is overwritten by `description`'s regular string.

If you wanted to `//` with an attrset so that `description` would be there for symmetry sake, `description = lib.mdDoc ''...` and `lib.mkEnableOption null // {...` would work.

But, I think this is a mistake because `package = lib.mkPackageOptionMD` is also asymmetrical with the rest of the `lib.mkOption`s.

Comment (fufexan) [reply nixos/modules/programs/regreet.nix]: I clearly needed sleep. I forgot to add `lib.mdDoc` before `''`. Couldn't figure out what was wrong. Thanks.

File: nixos/modules/programs/regreet.nix (hunk 4 above)
Comment (lilyinstarlight) [reply nixos/modules/programs/regreet.nix:55]: Wait did this get fixed? They still look reversed to me, but I may just be having a dumb moment

File: nixos/doc/manual/release-notes/rl-2305.section.md [hunk 7]
```diff
@@ -75,6 +75,8 @@ In addition to numerous new and upgraded packages, this release has the followin
 
 - [woodpecker-server](https://woodpecker-ci.org/), a simple CI engine with great extensibility. Available as [services.woodpecker-server](#opt-services.woodpecker-server.enable).
 
+- [ReGreet](https://github.com/rharish101/ReGreet), a clean and customizable greeter for greetd. Available as [programs.greetd](#opt-programs.regreet.enable).
```
[the remark this answers is not in the document]
Comment (fufexan) [reply nixos/doc/manual/release-notes/rl-2305.section.md:78]: Good catch.

File: nixos/modules/programs/regreet.nix [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+    environment.etc = {
+      "greetd/regreet.css".source =
+        if lib.isPath cfg.extraCss
+        then {source = cfg.extraCss;}
+        else {text = cfg.extraCss;};
+
+      "greetd/regreet.toml" =
+        if lib.isPath cfg.settings
+        then settingsFormat.generate "regreet.toml" cfg.settings
+        else cfg.settings;
```
Comment (lilyinstarlight) [inline nixos/modules/programs/regreet.nix:57]:
```suggestion
        if lib.isPath cfg.settings
        then cfg.settings
        else settingsFormat.generate "regreet.toml" cfg.settings;
```

Is this reversed between `cfg.settings` and `settingsFormat.generate "regreet.toml" cfg.settings`?

Comment (fufexan) [reply nixos/modules/programs/regreet.nix:57]: Ah, yeah.

Comment (fufexan) [reply nixos/modules/programs/regreet.nix]: Should have also been `"regreet.toml".source`. Fixed that as well.

File: nixos/modules/programs/regreet.nix [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+
+        To use ReGreet, {option}`services.greetd` has to be enabled and
+        {option}`services.greetd.settings.default_session` should contain the
+        appropriate configuration to launch
+        {option}`config.programs.regreet.package`. For examples, see the
+        [ReGreet Readme](https://github.com/rharish101/ReGreet#set-as-default-session).
+      '';
+    };
+
+    package = lib.mkPackageOptionMD pkgs "regreet" { };
```
Comment (lilyinstarlight) [inline nixos/modules/programs/regreet.nix:24]:
```suggestion
    package = lib.mkPackageOptionMD pkgs [ "greetd" "regreet" ] { };
```

This is only available as `pkgs.greetd.regreet`, right, not as `pkgs.regreet`?

Comment (fufexan) [reply nixos/modules/programs/regreet.nix:24]: Right.

Comment (lilyinstarlight) [reply nixos/modules/programs/regreet.nix:24]: Yeah then it needs that diff to eval

**Comment by lilyinstarlight** [comment]
So I just tested the module and it works with cage, but it is pretty awkward to have to do something else in another module just to get this module to do anything

Perhaps we can have it `lib.mkDefault` a `services.greetd.default_session.command` that just uses cage?

That way a user can set their own but it at least does something reasonable by default

**Comment by fufexan** [comment]
Sure, sounds good.

**Review by RaitoBezarius — CHANGES_REQUESTED** [review]
Untested; this looks good to me, I would ask though for an assertion regarding `services.greetd.enable` to protect the user.

**Comment by fufexan** [comment]
> albeit it was wonky with the fingerprint reader -- it made me click "Login" twice?

There's a fix coming up in https://github.com/rharish101/ReGreet/pull/4.

> Untested; this looks good to me, I would ask though for an assertion regarding `services.greetd.enable` to protect the user.

What should the assertion check? If we go Lily's route, the greetd service will be enabled by default, in which case I can only imagine adding a trace to let the user know greetd is enabled (but even that may not be necessary).

**Comment by RaitoBezarius** [comment]
> > Untested; this looks good to me, I would ask though for an assertion regarding `services.greetd.enable` to protect the user.
> 
> What should the assertion check? If we go Lily's route, the greetd service will be enabled by default, in which case I can only imagine adding a trace to let the user know greetd is enabled (but even that may not be necessary).

Whatever makes sense in the context. :)
Going Lily's route is a good idea I think.

File: nixos/modules/programs/regreet.nix [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+      description = lib.mdDoc ''
+        Enable ReGreet, a clean and customizable greeter for greetd.
+
+        To use ReGreet, {option}`services.greetd` has to be enabled and
+        {option}`services.greetd.settings.default_session` should contain the
+        appropriate configuration to launch
+        {option}`config.programs.regreet.package`. For examples, see the
+        [ReGreet Readme](https://github.com/rharish101/ReGreet#set-as-default-session).
+
+        A minimal configuration that launches ReGreet in `cage` is enabled by
```
Comment (lilyinstarlight) [inline nixos/modules/programs/regreet.nix:22]:
```suggestion
        A minimal configuration that launches ReGreet in {command}`cage` is enabled by
```

This is super nitpicky, but other than this, the PR looks good to me now!

**Review by lilyinstarlight — APPROVED** [review]
This looks good to me now! Did you have any other thoughts @spikespaz or @RaitoBezarius?

**Review by RaitoBezarius — APPROVED** [review]
Good to go.

**Comment by spikespaz** [comment]
> Untested; this looks good to me, I would ask though for an assertion regarding `services.greetd.enable` to protect the user.

> That way a user can set their own but it at least does something reasonable by default (assuming services.greetd.enable is turned on -- though maybe we can default-enable that too)

I think that enabling greetd should be entirely optional.

File: nixos/modules/programs/regreet.nix [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+        appropriate configuration to launch
+        {option}`config.programs.regreet.package`. For examples, see the
+        [ReGreet Readme](https://github.com/rharish101/ReGreet#set-as-default-session).
+
+        A minimal configuration that launches ReGreet in {command}`cage` is
+        enabled by this module by default.
+      '';
+    };
+
+    package = lib.mkPackageOptionMD pkgs [ "greetd" "regreet" ] { };
```
Comment (spikespaz) [inline nixos/modules/programs/regreet.nix:27]: Just curious, why did you add the `greetd` package here?

Comment (lilyinstarlight) [reply nixos/modules/programs/regreet.nix:27]: You have to use an array when the package is in a packageset (i.e. you normally do `pkgs.greetd.regreet`). If you do `"greetd.regreet"` instead of `[ "greetd" "regreet" ]` it will say the package is not found

**Comment by lilyinstarlight** [comment]
> I think that enabling greetd should be entirely optional.
> 
> Edit: It clearly already is. Disregard me. Looks good.

Heh, it happens (I was typing a response until I saw the edit)

Would adding a line in the docs saying you can explicitly opt-out with `services.greetd.enable = false` help?

**Comment by fufexan** [comment]
Updated to latest master. Anything else?

**Comment by spikespaz** [comment]
> Would adding a line in the docs saying you can explicitly opt-out with `services.greetd.enable = false` help?

Now that the default config is generated for greetd, and any newcomer to Nix should be reading the source code for modules, I think such instruction is unnecessary.

But I only say this because the scope of the module has shifted slightly, as it now provides a default configuration for `default_session`.
